### DISCUSSION TITLE
How to access project referenced assemblies in source generator?

### DISCUSSION THREAD

**Discussion by prabhav-mehra** [dotnet/roslyn#68281] [Q&A]
I have 3 projects.

Project 1: Simple Dotnet MAUI App.

Project 2: A Class Library which acts as a custom NuGet Package which does some work.

Project 3: A Class library which has a C# SourceGenerator. It creates a C# file from a xml file with property names in sync with keys in the xml file.

Now,

Project 1, has a reference to Project 2 using the NuGet Package, or local reference using ProjectReference.

Project 2, has a reference to Project 3, i.e. the SourceGenerator Project using ProjectReference.

The problem: 
The XML file that SourceGenerator uses is in Project 1. How can the SourceGenerator project get the location of that XML file from Project 1. I have set the build action to c# analyzer additional file and copy if newer so its in the output directory ( along with the app dlls).

So In theory,
Project 1 has XML file and has a reference to Project 2 -> Project 2 has a reference to Project 3 -> Project 3 uses that XML file from Project 1 and adds the property names as a cs file in Project 2.

Project 2 can't have that XML file, Since Multiple Projects can have a Reference to Project 2, that would mean the XML file content can be different.

I tried getting the path using Assembly locations but it gives the location to project 2 and not Project 1.  
Any help with how this can work or if there is any other well known way to handle such tasks.

Thanks.

**Comment by CyrusNajmabadi** [comment]
You cannot get to this data without referencing it directly. A compilation can only reference another compilation, not the things that compilation itself depended on.

**Reply by prabhav-mehra** [reply]
So the best bet is to maintain a .xml file in project 2?

is there a way to run a source generator on demand?
One solution i was thinking of was using Tasks in the csproj file
So have a task which sets the file path to the xml file in Project 1 ( core compile). Sends it to Project 2 using Additional Properties. Which then sends it to Project 3 (Source generator) and then only it runs?
Is this possible?

@CyrusNajmabadi

**Reply by prabhav-mehra** [reply]
I am new to the SourceGenerator world so a quick question:

Is it possible I can Reference the SourceGenerator Project in my Maui app, so the source generator can access the .xml file, But, make it create a class in Project 2 (Which also has a reference to the SourceGenerator project) instead.
@CyrusNajmabadi
